File: src/test/java/com/sfh/pokeRogueBot/service/ShortTermMemoryTest.java
package com.sfh.pokeRogueBot.service;

import com.sfh.pokeRogueBot.model.exception.ActionLoopDetectedException;
import com.sfh.pokeRogueBot.phase.impl.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.spy;

class ShortTermMemoryTest {

    ShortTermMemory shortTermMemory;
    int memorySize = 50;
    int uniquePhasesThreshold = 3;

    @BeforeEach
    void setUp() {
        ShortTermMemory objToSpy = new ShortTermMemory(memorySize, uniquePhasesThreshold);
        shortTermMemory = spy(objToSpy);
    }

    @Test
    void a_loop_is_detected(){

        // There is no loop check till the memory is full
        assertDoesNotThrow(() -> {
            for(int i = 0; i < memorySize / 2; i++) {
                shortTermMemory.memorizePhase(CommandPhase.NAME);
                shortTermMemory.memorizePhase(SwitchPhase.NAME);
            }
        });

        // if the first memory is added when the memory is full, the first check for loop is done
        assertThrows(ActionLoopDetectedException.class, () -> {
            shortTermMemory.memorizePhase(CommandPhase.NAME);
        });
    }

    @Test
    void no_loop_is_detected_if_different_phases_are_present(){
        assertDoesNotThrow(() -> {
            for(int i = 0; i < memorySize * 2; i++) {
                shortTermMemory.memorizePhase(CommandPhase.NAME);
                shortTermMemory.memorizePhase(SwitchPhase.NAME);
                shortTermMemory.memorizePhase(EggHatchPhase.NAME);
                shortTermMemory.memorizePhase(CheckSwitchPhase.NAME);
                shortTermMemory.memorizePhase(TitlePhase.NAME);
            }
        });
    }

    @Test
    void no_check_for_loop_should_happen_till_the_memory_is_full() {
        assertDoesNotThrow(() -> {
            for (int i = 0; i < memorySize - 1; i++) {
                shortTermMemory.memorizePhase(CommandPhase.NAME);
            }
        });
    }

    @Test
    void clearMemory_resets_the_memory(){
        assertDoesNotThrow(() -> {
            for (int i = 0; i < memorySize - 1; i++) {
                shortTermMemory.memorizePhase(CommandPhase.NAME);
            }
        });

        shortTermMemory.clearMemory();

        assertDoesNotThrow(() -> {
            for (int i = 0; i < memorySize - 1; i++) {
                shortTermMemory.memorizePhase(CommandPhase.NAME);
            }
        });
    }
}

File: src/test/kotlin/com/sfh/pokeRogueBot/service/ShortTermMemoryTest.kt
package com.sfh.pokeRogueBot.service

import com.sfh.pokeRogueBot.model.exception.ActionLoopDetectedException
import com.sfh.pokeRogueBot.phase.impl.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows

class ShortTermMemoryTest {

    private lateinit var shortTermMemory: ShortTermMemory
    private val memorySize = 50
    private val uniquePhasesThreshold = 3

    @BeforeEach
    fun setUp() {
        shortTermMemory = ShortTermMemory(memorySize, uniquePhasesThreshold)
    }

    @Test
    fun `a loop is detected`() {
        // There is no loop check till the memory is full
        assertDoesNotThrow {
            repeat(memorySize / 2) {
                shortTermMemory.memorizePhase(CommandPhase::class.simpleName!!)
                shortTermMemory.memorizePhase(SwitchPhase::class.simpleName!!)
            }
        }

        // if the first memory is added when the memory is full, the first check for loop is done
        assertThrows<ActionLoopDetectedException> {
            shortTermMemory.memorizePhase(CommandPhase::class.simpleName!!)
        }
    }

    @Test
    fun `no loop is detected if different phases are present`() {
        assertDoesNotThrow {
            repeat(memorySize * 2) {
                shortTermMemory.memorizePhase(CommandPhase::class.simpleName!!)
                shortTermMemory.memorizePhase(SwitchPhase::class.simpleName!!)
                shortTermMemory.memorizePhase(EggHatchPhase::class.simpleName!!)
                shortTermMemory.memorizePhase(CheckSwitchPhase::class.simpleName!!)
                shortTermMemory.memorizePhase(TitlePhase::class.simpleName!!)
            }
        }
    }

    @Test
    fun `no check for loop should happen till the memory is full`() {
        assertDoesNotThrow {
            repeat(memorySize - 1) {
                shortTermMemory.memorizePhase(CommandPhase::class.simpleName!!)
            }
        }
    }

    @Test
    fun `clearLastPhaseMemory resets the memory`() {
        assertDoesNotThrow {
            repeat(memorySize - 1) {
                shortTermMemory.memorizePhase(CommandPhase::class.simpleName!!)
            }
        }

        shortTermMemory.clearLastPhaseMemory()

        assertDoesNotThrow {
            repeat(memorySize - 1) {
                shortTermMemory.memorizePhase(CommandPhase::class.simpleName!!)
            }
        }
    }
}